bwt/wavelet.go

```diff
@@ -0,0 +1,356 @@
+package bwt
+
+import (
+	"fmt"
+	"math"
+
+	"golang.org/x/exp/slices"
+)
+
+/*
+
+For the waveletTree's usage, please read the its
+method documentation. To understand what it is and how
+it works for either curiosity or maintenance, then read below.
+
+# WaveletTree
+
+The Wavelet Tree allows us to conduct RSA queries on strings. in
+a memory and run time efficient manner.
+RSA stands for (R)ank, (S)elect, (A)ccess.
+
+## The Character's Path Encoding
+
+One important component is a character's path encoding.
+Which character we are working with in a given path in the tree.
+For example, given the alphabet A B C D E F G, a possible encoding is:
+
+A: 000
+B: 001
+C: 010
+D: 011
+E: 100
+F: 101
+G: 110
+H: 111
+
+If we wanted to get to the leaf that represent the character D, we'd
+take the path:
+
+   root
+  /
+left
+  \
+ right
+    \
+   right
+
+## The Data Represented at each node
+
+Let us consider the sequence "bananas"
+It has the alphabet b, a, n, s
+Let's say it has the encoding:
+a: 00
+n: 01
+b: 10
+s: 11
+and that 0 if left and 1 is right
+We can represent this tree with bitvectors:
+
+     0010101
+    /       \
+  1000      001
+ /    \    /   \
+a      n  b     s
+
+If we translate each bit vector to its corresponding string, then it becomes:
+
+     bananas
+    /       \
+  baaa      nns
+ /    \    /   \
+a      b  n     s
+
+## RSA Intuition
+
+From here you may be able to build some intuition as to how we can take RSA queries given
+a characters path encoding and which character we'd like to Rank, Select, and Access.
+
+### Rank Example
+
```

Comment:
```suggestion

WaveletTree.Rank(c, n) returns the rank of character c at index n in a sequence, i.e. how many
times c has occurred in a sequence before index n.

```

Just adding definition of rank because I didn't know it myself.